Dict expression memory pool

ampli opened the pull request:
Implement a dict expression memory pool.

Benefits:
1. Speedup.
1.1. The dict read/close is faster. It can be seen by running `tests.py`, which is dominated by dicts read/close, and runs about 6% faster.
2.2. The rest of the code also runs faster, maybe because the Exp memory is more packed and continuous, and the Exp struct is smaller (`next ` field removed).

A benchmark based on 3000 runs (the dict opening is a small part of the total CPU, especially for English, in which most of the speedup is from the rest of the code):
```
Parameters:   -n 600 -r 5
 1.726000 -   1.6990 =   -0.0270  +1.6%    data/en/corpus-basic.batch
 4.929000 -   4.9060 =   -0.0230  +0.5%    data/en/corpus-fixes.batch
 1.325000 -   1.2990 =   -0.0260  +2.0%    data/ru/corpus-basic.batch
```
2. It allows to add a pointer field in the Exp struct without exceeding 32 bytes (A WIP that eliminates `E_list` uses that).
3. It allows using common functions for expression handling for dict and rest of code, thus allowing code sharing. One of my WIPs concerns expression simplifier, and the function that worked on expression copy could not work on original dict expressions due to their different memory management. Several other functions are different due to historical reasons and still need a unification (TODO).

linas commented on link-grammar/api.c:
```diff
@@ -537,7 +537,11 @@ void sentence_delete(Sentence sent)
 	pool_delete(sent->Exp_pool);
```
above, you deleted what you're reusing!?

ampli replied:
As I couldn't find `pool_reuse(sent->Exp_pool)` in the code, I guess you referred to line 543:
`pool_reuse(sent->dict->Exp_pool)`.
But this is the dict Exp memory-pool.

BTW:
1. **All** the sentence memory pools are currently deleted at `sentence_delete()`.
But the dict pools survive - for a file dict they are just left untouched, and for a db dict they are reused.
2. I guess it is possible to save CPU by reusing memory-pools over sentences and deleting them on thread exit only.
This can be done by reusing them on `sentence_delete()` instead of deleting them, and deleting them using a registered exit function (I didn't test that so I don't know how much it would save).
A drawback is that if a single sentence happened to use a lot of memory, this memory would not get freed until exit.
I remedy could be adding an additional argument to `pool_reuse()` for how many memory-pool entries to retain.

linas commented:
Thanks.

> A WIP that eliminates E_list

Be aware that this is currently a part of the external API; tehre's only one user of that API, and it can be changed as needed. As long as its possible to have access to Exp's or anything like them, its OK.

ampli commented:
> Be aware that this is currently a part of the external API; tehre's only one user of that API, and it can be changed as needed. As long as its possible to have access to Exp's or anything like them, its OK.

I did the change in a compatible way - my new `Exp_struct` is actually what used to be the E_list:
The eliminated `E_list struct` includes a `next` pointer (for the next operand) and an Exp pointer.
Instead of the Exp pointer I used the full Exp struct, and renamed the said `next` pointer to `operand_next` as follows:
```
     46 struct Exp_struct
     47 {
     48         Exp *operand_next; /* Next same-level operand. */
     49         Exp_type type; /* One of three types: AND, OR, or connector. */
     50         char dir;      /* The connector connects to the left ('-') or right ('+'). */
     51         bool multi;    /* TRUE if a multi-connector (for connector). */
     52         union
     53         {
     54                 Exp *operand_first; /* First operand (for non-terminals). */
     55                 condesc_t *condesc; /* Only needed if it's a connector. */
     56         };
     57         double cost;   /* The cost of using this expression. */
     58 };
```
(Note that the original `next` pointer of `Exp_struct`, used only for memory management, got eliminated in the the recent commit bae7d591e that introduced a dict Exp memory pool).

For the reference, this is the eliminated `E_list_struct`:
```
struct E_list_struct  // renamed to Exp_struct
{
        E_list * next;    // renamed to operand_next
        Exp * e;           // the full previous Exp_struct is used instead
};
```
In addition, for readability I renamed `u.l` to `operand_first` (seems clearer than `operand_list` due to `operand_next`).

So the code conversion is rather straightforward, but in the LG lib most of the specific E_list_struct processing code got eliminated.

**Proposal:**
Add additional API call to make Exp future proof to internal changes (that may be needed according to my WIPs and TODOs):
```
Exp *lg_exp_operand_first(Exp *);
Exp *lg_exp_operand_next(Exp *);
```
When Exp is an opaque structure pointer.

ampli commented:
Discussion on eliminating E_list moved to #718.
